Support List[str].index (#39210)

firstprayer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40348 Support List[str].index (#39210)**

Differential Revision: [D22757035](https://our.internmc.facebook.com/intern/diff/D22757035)

Commit:
2nd approach for #39210

[ghstack-poisoned]

Commit:
Update on "2nd approach for #39210"




[ghstack-poisoned]

wanchaol commented on aten/src/ATen/core/List_inl.h:
```diff
@@ -129,6 +129,12 @@ template<class T, class Iterator>
+
+template<class T, class Iterator>
+bool operator==(const ListElementReference<T, Iterator>& lhs, const T& rhs) {
+  T lhs_tmp = lhs;
```
this overload looks fine to me. cc @smessmer will this custom == overload have any implications on the list impl?

smessmer replied:
Let's restrict it to only work on `const ListElementReference<T, Iterator>& rhs`. `operator==` should be symmetric or the API gets weird because `a == b` does something different than `b == a`.

firstprayer replied:
@smessmer just to confirm, is the suggestion to add a mirror version of this overload as well, or just not have overload at all? 

If without this overload, then std::find() will not compile (see https://github.com/pytorch/pytorch/pull/40327#discussion_r443090950)

firstprayer replied:
> @smessmer just to confirm, is the suggestion to add a mirror version of this overload as well, or just not have overload at all?
> 
> If without this overload, then std::find() will not compile (see [#40327 (comment)](https://github.com/pytorch/pytorch/pull/40327#discussion_r443090950))

@smessmer ping :)

smessmer replied:
ok sg. I meant to remove the overload, but `std::find` convinced me. Having it symmetrical is fine.

firstprayer commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -881,6 +882,10 @@ RegisterOperators reg2({
         "aten::eq.bool_list(bool[] a, bool[] b) -> bool",
         listEq<bool>,
         aliasAnalysisFromSchema()),
+    Operator(
```
I think it's covered in one the new test case added (the test_remove one), do we need a dedicated test for this?

wanchaol replied:
No need to add a separate test case, can you rename it to `test_str_list_remove_and_eq` to indicate we are testing both?

wanchaol reviewed (commented):
Looks good, could you either redirect this change to the first PR, or rename this PR with title and proper description so that we can merge it? thanks!

Commit:
Update on "Support List[str].index (#39210)"




[ghstack-poisoned]

firstprayer reviewed (commented):
added the symmetry overload

Commit:
Update on "Support List[str].index (#39210)"




[ghstack-poisoned]

Commit:
Update on "Support List[str].index (#39210)"


Differential Revision: [D22757035](https://our.internmc.facebook.com/intern/diff/D22757035)

[ghstack-poisoned]

wanchaol reviewed (approved):
looks good, thanks!
